From d3494ad1c9e7f4f91ae7e003c5e855dc880bc5f3 Mon Sep 17 00:00:00 2001
From: Marcelo de Castro Loebens <Marcelo.DeCastroLoebens@windriver.com>
Date: Fri, 4 Sep 2026 15:41:15 -0400
Subject: [PATCH] Add tls-ignore-hostname remote option

Add a boolean option "tls-ignore-hostname" (default false). When true,
the certificate hostname/SAN check is skipped (CURLOPT_SSL_VERIFYHOST=0)
while the CA chain is still verified (unless tls-permissive is also set).

On Bullseye, tls-permissive would skip verification in both CA chain and
hostname/SANs. On Trixie, tls-permissive only skips the verification of
the CA chain, but still verifies hostname/IPs in the server certificate,
causing a change of behaviour.

Signed-off-by: Marcelo de Castro Loebens <Marcelo.DeCastroLoebens@windriver.com>
---
 man/ostree.repo-config.xml          |  8 ++++++++
 src/libostree/ostree-fetcher-curl.c |  6 ++++++
 src/libostree/ostree-fetcher.h      |  1 +
 src/libostree/ostree-repo-pull.c    | 11 +++++++++++
 4 files changed, 26 insertions(+)

diff --git a/man/ostree.repo-config.xml b/man/ostree.repo-config.xml
index 181d7ba..1183317 100644
--- a/man/ostree.repo-config.xml
+++ b/man/ostree.repo-config.xml
@@ -328,6 +328,14 @@ License along with this library. If not, see <https://www.gnu.org/licenses/>.
         certificate will be accepted.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>tls-ignore-hostname</varname></term>
+        <listitem><para>A boolean value, defaults to false.  If set to
+        true, the server TLS certificate hostname (Subject Alternative
+        Name / Common Name) is not checked against the host used to
+        connect.</para></listitem>
+      </varlistentry>
+
       <varlistentry>
         <term><varname>tls-client-cert-path</varname></term>
         <listitem><para>Path to file for client-side certificate, to present when making requests to this repository.</para></listitem>
diff --git a/src/libostree/ostree-fetcher-curl.c b/src/libostree/ostree-fetcher-curl.c
index 5c3bdd9..a960956 100644
--- a/src/libostree/ostree-fetcher-curl.c
+++ b/src/libostree/ostree-fetcher-curl.c
@@ -884,6 +884,12 @@ initiate_next_curl_request (FetcherRequest *req, GTask *task)
       g_assert_cmpint (rc, ==, CURLM_OK);
     }
 
+  if ((self->config_flags & OSTREE_FETCHER_FLAGS_TLS_NO_VERIFY_HOSTNAME) > 0)
+    {
+      rc = curl_easy_setopt (req->easy, CURLOPT_SSL_VERIFYHOST, 0L);
+      g_assert_cmpint (rc, ==, CURLM_OK);
+    }
+
   if (self->tls_client_cert_path)
     {
       /* Support for pkcs11:
diff --git a/src/libostree/ostree-fetcher.h b/src/libostree/ostree-fetcher.h
index 59ae633..ae8273c 100644
--- a/src/libostree/ostree-fetcher.h
+++ b/src/libostree/ostree-fetcher.h
@@ -55,6 +55,7 @@ typedef enum
   OSTREE_FETCHER_FLAGS_TLS_PERMISSIVE = (1 << 0),
   OSTREE_FETCHER_FLAGS_TRANSFER_GZIP = (1 << 1),
   OSTREE_FETCHER_FLAGS_DISABLE_HTTP2 = (1 << 2),
+  OSTREE_FETCHER_FLAGS_TLS_NO_VERIFY_HOSTNAME = (1 << 3),
 } OstreeFetcherConfigFlags;
 
 typedef enum
diff --git a/src/libostree/ostree-repo-pull.c b/src/libostree/ostree-repo-pull.c
index 9989415..e28d29e 100644
--- a/src/libostree/ostree-repo-pull.c
+++ b/src/libostree/ostree-repo-pull.c
@@ -2942,6 +2942,7 @@ _ostree_repo_remote_new_fetcher (OstreeRepo *self, const char *remote_name, gboo
   OstreeFetcher *fetcher = NULL;
   OstreeFetcherConfigFlags fetcher_flags = 0;
   gboolean tls_permissive = FALSE;
+  gboolean tls_ignore_hostname = FALSE;
   OstreeFetcherSecurityState ret_state = OSTREE_FETCHER_SECURITY_STATE_TLS;
   gboolean success = FALSE;
 
@@ -2958,6 +2959,16 @@ _ostree_repo_remote_new_fetcher (OstreeRepo *self, const char *remote_name, gboo
       ret_state = OSTREE_FETCHER_SECURITY_STATE_INSECURE;
     }
 
+  if (!ostree_repo_get_remote_boolean_option (self, remote_name, "tls-ignore-hostname", FALSE,
+                                              &tls_ignore_hostname, error))
+    goto out;
+
+  if (tls_ignore_hostname)
+    {
+      fetcher_flags |= OSTREE_FETCHER_FLAGS_TLS_NO_VERIFY_HOSTNAME;
+      ret_state = OSTREE_FETCHER_SECURITY_STATE_INSECURE;
+    }
+
   if (gzip)
     fetcher_flags |= OSTREE_FETCHER_FLAGS_TRANSFER_GZIP;
 
-- 
2.43.0

